### PR TITLE
Redo force spreading 1

### PR DESCRIPTION
**Pull request by drwells**
Most of #737 - I will address the parts pertinent to FEDataManager in another PR.

First part of #830 - now that we have the forces spread onto their own index we can go ahead and redo the way we accumulate data in parallel without worrying about other forces.

I encountered a few bugs along the way while writing this patch - to aid in future debugging I added a new test and modified an existing test so that we have an IBFE-based simulation that executes multiple regrids and gives the same answers with and without the scratch hierarchy.

### CODE REVIEW

File: src/IB/IBFEMethod.cpp [hunk 1]
```diff
@@ -929,33 +929,67 @@ IBFEMethod::spreadForce(const int f_data_idx,
...
+        // harder case: we have a scratch index on the scratch hierarchy but
+        // no corresponding index on the primary hierarchy.
+        //
+        // unlike the other data ops, this is always on the primary hierarchy
+        auto f_primary_data_ops = HierarchyDataOpsManager<NDIM>::getManager()->getOperationsDouble(f_var, d_hierarchy);
+        f_primary_data_ops->resetLevels(ln, ln);
+        const auto f_primary_scratch_data_idx = d_primary_eulerian_data_cache->getCachedPatchDataIndex(f_data_idx);
+        // we have to zero everything here since the scratch to primary
+        // communication does not touch ghost cells, which may have junk
+        f_primary_data_ops->setToScalar(f_primary_scratch_data_idx, 0.0, /*interior_only*/ false);
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp]: @boyceg this is what I was missing an hour ago - oops.

**Comment by drwells** [comment]
@boyceg could you take a quick look at this soon? I need this PR merged to proceed with the force spreading redo.

File: src/IB/IBFEMethod.cpp (hunk 1 above)
Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: OK, so accumulation from the ghost region to the interior is done separately for each structure?

**Comment by drwells** [comment]
We have a test based on ex8 - I should check again but I believe those structures are right at the boundary.

It would be wise to add some more spreading tests that explicitly test spreading when we have a structure on the boundary - I'll add that to the relevant issue.

File: src/IB/IBFEMethod.cpp (hunk 1 above)
Comment (drwells) [reply src/IB/IBFEMethod.cpp]: No - see circa line 950. We do the accumulation after looping over all parts.

**Comment by boyceg** [comment]
The thing I would suggest is to have a case that has two structures that are close to each other that are right at the boundary. They even could be clones of each other.

File: src/IB/IBFEMethod.cpp (hunk 1 above)
Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: Do different parts have different scratch hierarchies?

Comment (drwells) [reply src/IB/IBFEMethod.cpp]: No, but they do presently use different scratch patch data indices - this PR was written before the overloads #858 were added to avoid using a scratch force data index inside `FEDataManager::spread`.

Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: Can you remind me --- is there one scratch hierarchy, or are there separate scratch hierarchies for each part?

Comment (boyceg) [reply src/IB/IBFEMethod.cpp]: (As you can probably guess, I am worried about losing the ghost values.)

Comment (drwells) [reply src/IB/IBFEMethod.cpp]: There is just one scratch hierarchy managed by IBFEMethod. The FEDataManagers in `d_scratch_fe_data_managers` all store `Pointer`s to that hierarchy instead of the 'normal' one.

Comment (drwells) [reply src/IB/IBFEMethod.cpp]: That's a really good point. Even though we have some tests based on `ex8` in the test suite which this passes, lets be paranoid and add some more spreading tests (as described below) before proceeding.

**Comment by drwells** [comment]
I fixed the merge conflict on my own machine but now the `ex8` tests (an IBFE example with a structure on the boundary) fail. At least we know that the test suite is good enough to find a bug with this implementation :)

**Comment by drwells** [comment]
I found the problem - we don't properly handle ghost data in the overload of FEDataManager::spread() that takes an optional boundary op object and the `close` booleans. The new version of this PR works because it uses a different overload.

I'll put up another PR that fixes that bug in a minute.

**Comment by drwells** [comment]
@boyceg you approved this 10 days ago - the changes on top of those are minor (its just the test addition and the use of a different `FEDataManager::spread()` overload). Would you please take a quick look at this when you have a minute?
